### PR TITLE
feat(apollo-router-plugin): persisted operations (app deployments) plugin with Hive, update usage to API-Version v2

### PR DESCRIPTION
**Pull request by dotansimha**
closes https://github.com/graphql-hive/platform/issues/5498 

- [x] Add README for the library page
- [x] Implement persisted operations plugin
- [x] Testing (code)
- [x] Testing (real Hive project) 
- [x] Send hash along with usage reporting
- [x] Update usage reporting to use v2
- [x] Add more unit tests to usage reporting v2
- [x] fix TODOs and clean up the code
- [x] Documentation
- [x] Product update

### CODE REVIEW

File: packages/web/docs/src/components/usage-reports-json-schema.tsx
Comment (n1ru4l) [inline packages/web/docs/src/components/usage-reports-json-schema.tsx:32]: 🫶

**Comment by YassinEldeeb** [comment]
will review later today!

File: packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx:2]:
```suggestion
title: App Deployments Support for Apollo Router
```

File: packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx:256]: missing link here?

File: packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx:254]: I would probably use Persisted Documents as the main title here 🤔

**Review by n1ru4l — APPROVED** [review]

File: packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx:270]:
```suggestion
to send persisted document requests from your app, using the Hive CDN for resolving the
```

File: packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx:268]:
```suggestion
Once enabled, follow the
```

File: packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/docs/other-integrations/apollo-router.mdx:312]: I assume these are related to resolving the documents from the CDN? If so it should probably be noted there.

**Review by saihaj — APPROVED** [review]

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:313]: I think there should be a retry mechanism for CDN fetching, it only tries once here.

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:330]:
```suggestion
        warn!(
            "Document fetch from CDN failed: HTTP {}, Body: {:?}",
            response.status(),
            response.text().await.unwrap_or_else(|_| "Unavailable".to_string())
        );
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:284]: there's no differentiation between a malformed JSON body and a missing document ID, maybe there should be more variants for clearer error messages:
- `DOCUMENT_ID_MISSING`
- `MALFORMED_REQUEST_BODY`

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:335]: I'm not sure how often the CDN's persisted queries are updated, but maybe when we face the case where we can't find the document in the CDN, we should add a short-lived cache (maybe with 5min TTL), so that subsequent requests for the same non-found document don't repeat the CDN fetch, overloading it unnecessarily.

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:329]: this will likely reduce the throughput of the incoming requests to the gateway, bc we're locking the cache on every successful request, which means that if the gateway can process 10 requests at once (concurrently), this will be a bottleneck because each request will wait for the other to release the `cache` lock for it to be acquired by the other thread.

maybe `RwLock` or `dashmap` can be used for the cache instead of a `Mutex`?

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:506]: simplify with:

```rust
..Default::default()
```

to add value to the rest of the things that need a default value.

Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:506]: simplify with:

```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:537]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:574]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:610]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:646]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:685]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:719]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:749]:
```suggestion
            ..Default::default()
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (YassinEldeeb) [inline packages/libraries/router/src/persisted_documents.rs:318]: graceful error handling:

```suggestion
                            let document = response.text().await.map_err(PersistedDocumentsError::FailedToReadBody)?;
```

File: packages/libraries/router/src/persisted_documents.rs
Comment (dotansimha) [reply packages/libraries/router/src/persisted_documents.rs:336]: Hmm, good point, but this one is tricky, since we'll have to get `rwLock.write()` also for `cache.get` calls, because the `LruCache` needs `&mut self` in order to run `.get` calls (so write access is required, to update the LRU cache).

File: packages/libraries/router/src/persisted_documents.rs
Comment (dotansimha) [reply packages/libraries/router/src/persisted_documents.rs:325]: good point, fixed

File: packages/libraries/router/src/persisted_documents.rs
Comment (dotansimha) [reply packages/libraries/router/src/persisted_documents.rs:294]: Actually, the `structure ExpectedBodyStructure` is defined with `document_id: Option<String>` field, so it will not fail in case of a missing document id. 
The document ID is checked later for its value and then a specific error is returned. 
So this one can actually fail only in a case the json parsing fails.

File: packages/libraries/router/src/persisted_documents.rs
Comment (dotansimha) [reply packages/libraries/router/src/persisted_documents.rs]: See https://stackoverflow.com/questions/69617603/using-rwlock-read-on-a-mutable-struct

Comment (YassinEldeeb) [reply packages/libraries/router/src/persisted_documents.rs]: hmmm...you're right it's not as straightforward, let me see what I can do, I'll experiment with it

**Review by kamilkisiela — APPROVED** [review]

File: .changeset/rotten-islands-joke.md
Comment (n1ru4l) [inline .changeset/rotten-islands-joke.md:5]:
```suggestion
Add support for persisted documents using Hive App Deployments. [Learn more](link to docs)
```

File: packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx:1]: We need to update the date in the filename

File: packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx
Comment (n1ru4l) [inline packages/web/docs/src/pages/product-updates/2024-11-07-apollo-router-app-deployments/index.mdx:2]:
```suggestion
title: Persisted Documents Support for Apollo Router
```

**Pull request merged**
